utils/env_preferences/src/apple.rs

```diff
@@ -0,0 +1,167 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core_foundation_sys::{
+    array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+    base::{CFIndex, CFRelease, CFRetain},
+    calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+    locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+    string::{
+        kCFStringEncodingUTF8, CFStringGetCString, CFStringGetCStringPtr, CFStringGetLength,
+        CFStringRef,
+    },
+    timezone,
+};
+use libc::c_char;
+use std::ffi::{CStr, CString};
+
+use crate::RetrievalError;
+
+/// Helps to get string, when a string cannot be obtained in O(1) time and space directly from the pointer
+/// A buffer is created of size `length + 1`. After copying the string into the buffer, the null characters `\0`
+/// are removed from the string, which helps it converting to icu4x locale in tests
+fn get_string(ptr: CFStringRef) -> Result<String, RetrievalError> {
+    // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
+    // Returns pointer in O(1) without any memory allocation. This can return NULL so we are handling it by directly
+    // copying it using `CFStringGetCString`
+    let lang_str: *const c_char = unsafe { CFStringGetCStringPtr(ptr, kCFStringEncodingUTF8) };
+
+    if !lang_str.is_null() {
+        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }.to_str()?;
+        Ok(lang_rust_str.to_string())
+    } else {
+        // `lang_str` is null, i.e. `CFStringGetCStringPtr` couldn't give desired output, trying with
+        // manual allocations
+        // SAFETY: It returns length of the string, from above conditional statement we ensure
+        // that the `lang_ptr` is not NULL thus making it safe to call
+        let length = unsafe { CFStringGetLength(ptr) as usize };
+
+        let mut c_str_buf: Vec<u8> = vec![0; length + 1];
+
+        // SAFETY: Safety is ensured by following points
+        // 1. `lang_ptr` is not NULL, checked through conditional statement
+        // 2. `c_str_buf` is large enough and in scope after this call
+        unsafe {
+            CFStringGetCString(
+                ptr,
+                c_str_buf.as_mut_ptr() as *mut c_char,
+                c_str_buf.len() as CFIndex,
+                kCFStringEncodingUTF8,
+            );
+        }
+
+        let c_string = CString::from_vec_with_nul(c_str_buf)?;
+        let str_converted = c_string.into_string()?;
+        Ok(str_converted
```

Comment:
`CString::into_string` doesn't include the trailing `NUL`, so you shouldn't do this (it's an extra allocation).